File: execute_function.js
const { exec } = require('child_process');

function isDockerAvailable(callback) {
    exec('docker --version', (error, stdout, stderr) => {
        if (error) {
            console.error('❌ Error: Docker CLI not found. Make sure Docker is installed and accessible.');
            callback(false);
        } else {
            callback(true);
        }
    });
}

function executeFunction(functionName, lang, timeout = 5) {
    const imageName = `$javascript-function`; // Match built images
    console.log(`🚀 Running ${functionName} in ${lang} with a timeout of ${timeout} seconds...`);

    isDockerAvailable((available) => {
        if (!available) {
            process.exit(1);
        }

        const command = `docker run --rm ${imageName}`;
        const process = exec(command, { timeout: timeout * 1000 }, (error, stdout, stderr) => {
            if (error) {
                if (error.killed) {
                    console.error(`❌ Function ${functionName} exceeded ${timeout} seconds and was terminated.`);
                } else {
                    console.error(`❌ Error: ${stderr}`);
                }
            } else {
                console.log(`✅ Output: ${stdout}`);
            }
        });

        process.on('error', (err) => {
            console.error(`❌ Error: ${err.message}`);
        });
    });
}

if (process.argv.length < 4) {
    console.log('Usage: node execute_function.js <function_name> <language> [timeout]');
    process.exit(1);
}

const functionName = process.argv[2];
const lang = process.argv[3];
const timeout = process.argv[4] ? parseInt(process.argv[4], 10) : 5; // Default timeout: 5 seconds

executeFunction(functionName, lang, timeout);

File: execute_function.py
import subprocess
import sys
import time

def execute_function(function_name, lang, timeout=5):  
    """Runs a function inside a Docker container with a timeout."""
    
    image_name = f"example_function_{lang}"  # Selects the correct image
    print(f"🚀 Running {function_name} in {lang} with a timeout of {timeout} seconds...")

    try:
        # Start the container
        process = subprocess.Popen(
            ["docker", "run", "--rm", image_name],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )

        # Wait for execution with timeout
        start_time = time.time()
        while process.poll() is None:  # Process is still running
            if time.time() - start_time > timeout:
                print(f"⏳ Timeout exceeded! Stopping {function_name}...")
                process.terminate()  # Graceful stop
                time.sleep(2)  # Give some time to terminate
                process.kill()  # Force kill if still running
                raise TimeoutError(f"Function {function_name} exceeded {timeout} seconds!")

        # Get output
        output, error = process.communicate()
        if output:
            print(f"✅ Output: {output.decode().strip()}")
        if error:
            print(f"⚠ Error: {error.decode().strip()}")

    except TimeoutError as e:
        print(f"❌ {e}")

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python execute_function.py <function_name> <language> [timeout]")
        sys.exit(1)

    function_name = sys.argv[1]
    lang = sys.argv[2]
    timeout = int(sys.argv[3]) if len(sys.argv) > 3 else 5  # Default timeout: 5 seconds

    execute_function(function_name, lang, timeout)